File: practice/algorithms/strings/weighted_uniform_strings/WeightedUniformStrings.java
import java.util.ArrayList;
import java.util.List;
import java.util.Scanner;
import java.util.Set;
import java.util.stream.Collectors;
import java.util.stream.Stream;

public class WeightedUniformStrings {
    public static void main(String[] args) {
        Scanner stdin = new Scanner(System.in);
        String line = stdin.nextLine();
        List<Integer> queries = new ArrayList<>();
        while (stdin.hasNext()) {
            int q = Integer.parseInt(stdin.nextLine());
            queries.add(q);
        }
        queries = queries.subList(1, queries.size());
        List<String> subs = uniformSubstrings(line, new ArrayList<>());
        Set<Integer> substringWeights = subs.stream().flatMap(s -> {
            int c = s.charAt(0) - 'a' + 1;
            List<Integer> lengts = new ArrayList<>();
            for(int i = 1; i <= s.length(); i++) {
                lengts.add(i * c);
            }
            return lengts.stream();
        }).collect(Collectors.toSet());

        for(int q : queries) {
            if(substringWeights.contains(q)) {
                System.out.println("Yes");
            } else {
                System.out.println("No");
            }
        }
    }

    private static List<String> uniformSubstrings(String s, List<String> strings) {
        if(s.length() > 0) {
            char c = s.charAt(0);
            int substringIndex = takeWhile(s, c);
            String substring = s.substring(0, substringIndex);
            strings.add(substring);
            return uniformSubstrings(s.substring(substringIndex), strings);
        }
        return strings;
    }

    private static int takeWhile(String s, char c) {
        for(int i = 0; i < s.length(); i++) {
            if(s.charAt(i) != c) {
                return i;
            }
        }
        return s.length();
    }
}

File: practice/algorithms/strings/weighted_uniform_strings/WeightedUniformStrings.js
'use strict';

Array.prototype.flatMap = function(lambda) {
    return Array.prototype.concat.apply([], this.map(lambda));
};

const processData = input => {
    const lines = input.split('\n');
    const line = lines[0];
    const queries = lines.slice(2).map(s => parseInt(s));

    const subs = uniformSubstrings(line, []);
    const substringWeights = subs.flatMap(s => {
        const c = s.charCodeAt(0) - 'a'.charCodeAt(0) + 1;
        const lengts = [];
        for(let i = 1; i <= s.length; i++) {
            lengts.push(i * c);
        }
        return lengts;
    });

    for(let q = 0; q < queries.length; q++) {
        const query = queries[q];
        if(substringWeights.indexOf(query) != -1) {
            console.log("Yes");
        } else {
            console.log("No");
        }
    }
};

const uniformSubstrings = (s, strings) => {
    if(s.length > 0) {
        const c = s.charAt(0);
        const substringIndex = takeWhile(s, c);
        const substring = s.slice(0, substringIndex);
        const substring2 = s.substring(substringIndex);
        strings.push(substring);
        return uniformSubstrings(substring2, strings);
    }
    return strings;
}

const takeWhile = (s, c) => {
    for(let i = 0; i < s.length; i++) {
        if(s.charAt(i) !== c) {
            return i;
        }
    }
    return s.length;
}

process.stdin.resume();
process.stdin.setEncoding("ascii");

let _input = "";
process.stdin.on("data", input => _input += input);
process.stdin.on("end", () => processData(_input));